### ISSUE TITLE
N+1 Issue with `Decimal` data type and combining queries (batching)

### ISSUE THREAD

**Issue by mikeburgh** [prisma/prisma#5952]
## Bug description

When using Decimal for the data type, Prisma correctly combines the queries, however they are not returned correctly (un combined?) and graphQL throws an error (if set to notNull) or displays no data

## How to reproduce

The following repo has a reproducible example setup as a bare minimum with GraphQL and SQLite DB. 
Also provides notes on changing the schema to Int and re running to show it working correctly.

Clone https://github.com/mikeburgh/prisma-nexus and follow instructions in readme

## Expected behavior

Correct results to be returned once queries are combined

## Prisma information

```graphql
generator client {
  provider        = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model Parent {
  id      Decimal @id 
  childId Decimal 
  child   Child   @relation(fields: [childId], references: [id])
}

model Child {
  id     Decimal  @id 
  Parent Parent[]
}
```

See repo above for full details

## Environment & setup

- OS: Ubuntu
- Database: SQLite and Sql Server
- Node.js version: v14.15.4
- Prisma version:

```
prisma               : 2.18.0
@prisma/client       : 2.18.0
Current platform     : debian-openssl-1.1.x
Query Engine         : query-engine da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine     : migration-engine-cli da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine : introspection-core da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary        : prisma-fmt da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Studio               : 0.356.0
```

**Comment by tarazena** [comment]
@mikeburgh I don't think this is N+1 Issue, The reason your GraphQL server is not working on `Decimal` is because Prisma is now generating [Decimal.js](https://mikemcl.github.io/decimal.js/) Objects and not Numbers, Since Nexus does not know what `Decimal` is (Only `Float` is defined in nexus schema), the GraphQL server will throw an error because it is expecting a Float, thus the error.

To overcome this error, you can either:
- Define Decimal as part of the GraphQL schema  (maybe using [this](https://www.npmjs.com/package/graphql-type-decimal).
- Convert the `Decimal` to number like
```ts
resolve: async (parent, args, ctx) => {
                                 const r = ctx.prisma.child.findUnique({
					where: {
						id: parent.childId,
					}
                                  });
				return {...r, id: r.id.toNumber()}
				
			},
``` 
- Switch the `Decimal` to a different data types. 

I opened a [discussion post](https://github.com/prisma/prisma/discussions/5980) to maybe disable the `Decimal` generation type.

**Comment by mikeburgh** [comment]
@tarazena thanks for the insight.. but in the resolve method.. r is coming back as null, so there is no data to send back to nexus. 

Prisma is being cleaver when it gets the batch of queries for the children (to solve the N+1 issue) and grouping them up for the children: 

```
prisma:query SELECT `main`.`Parent`.`id`, `main`.`Parent`.`childId` FROM `main`.`Parent` WHERE 1=1 LIMIT ? OFFSET ?
  prisma:client Prisma Client call: +46ms
  prisma:client prisma.child.findUnique({
  prisma:client   where: {
  prisma:client     id: {
  prisma:client       constructor: function Decimal2(v) {
  prisma:client     var e, i2, t, x = this;
  prisma:client     if (!(x instanceof Decimal2))
  prisma:client       return new Decimal2(v);
  prisma:client     x.constructor = Decimal2;
  prisma:client     if (v instanceof Decimal2) {
  prisma:client       x.s = v.s;
  prisma:client       if (external) {
  prisma:client         if (!v.d || v.e > Decimal2.maxE) {
  prisma:client           x.e = NaN;
  prisma:client           x.d = null;
  prisma:client         } else if (v.e < Decimal2.minE) {
  prisma:client           x.e = 0;
  prisma:client           x.d = [0];
  prisma:client         } else {
  prisma:client           x.e = v.e;
  prisma:client           x.d = v.d.slice();
  prisma:client         }
  prisma:client       } else {
  prisma:client         x.e = v.e;
  prisma:client         x.d = v.d ? v.d.slice() : v.d;
  prisma:client       }
  prisma:client       return;
  prisma:client     }
  prisma:client     t = typeof v;
  prisma:client     if (t === "number") {
  prisma:client       if (v === 0) {
  prisma:client         x.s = 1 / v < 0 ? -1 : 1;
  prisma:client         x.e = 0;
  prisma:client         x.d = [0];
  prisma:client         return;
  prisma:client       }
  prisma:client       if (v < 0) {
  prisma:client         v = -v;
  prisma:client         x.s = -1;
  prisma:client       } else {
  prisma:client         x.s = 1;
  prisma:client       }
  prisma:client       if (v === ~~v && v < 1e7) {
  prisma:client         for (e = 0, i2 = v; i2 >= 10; i2 /= 10)
  prisma:client           e++;
  prisma:client         if (external) {
  prisma:client           if (e > Decimal2.maxE) {
  prisma:client             x.e = NaN;
  prisma:client             x.d = null;
  prisma:client           } else if (e < Decimal2.minE) {
  prisma:client             x.e = 0;
  prisma:client             x.d = [0];
  prisma:client           } else {
  prisma:client             x.e = e;
  prisma:client             x.d = [v];
  prisma:client           }
  prisma:client         } else {
  prisma:client           x.e = e;
  prisma:client           x.d = [v];
  prisma:client         }
  prisma:client         return;
  prisma:client       } else if (v * 0 !== 0) {
  prisma:client         if (!v)
  prisma:client           x.s = NaN;
  prisma:client         x.e = NaN;
  prisma:client         x.d = null;
  prisma:client         return;
  prisma:client       }
  prisma:client       return parseDecimal(x, v.toString());
  prisma:client     } else if (t !== "string") {
  prisma:client       throw Error(invalidArgument + v);
  prisma:client     }
  prisma:client     if ((i2 = v.charCodeAt(0)) === 45) {
  prisma:client       v = v.slice(1);
  prisma:client       x.s = -1;
  prisma:client     } else {
  prisma:client       if (i2 === 43)
  prisma:client         v = v.slice(1);
  prisma:client       x.s = 1;
  prisma:client     }
  prisma:client     return isDecimal.test(v) ? parseDecimal(x, v) : parseOther(x, v);
  prisma:client   },
  prisma:client       s: 1,
  prisma:client       e: 3,
  prisma:client       d: [
  prisma:client         1111
  prisma:client       ]
  prisma:client     }
  prisma:client   }
  prisma:client }) +0ms
  prisma:client Generated request: +0ms
  prisma:client query {
  prisma:client   findUniqueChild(where: {
  prisma:client     id: 1111
  prisma:client   }) {
  prisma:client     id
  prisma:client   }
  prisma:client }
  prisma:client  +0ms
  prisma:client Prisma Client call: +1ms
  prisma:client prisma.child.findUnique({
  prisma:client   where: {
  prisma:client     id: {
  prisma:client       constructor: function Decimal2(v) {
  prisma:client     var e, i2, t, x = this;
  prisma:client     if (!(x instanceof Decimal2))
  prisma:client       return new Decimal2(v);
  prisma:client     x.constructor = Decimal2;
  prisma:client     if (v instanceof Decimal2) {
  prisma:client       x.s = v.s;
  prisma:client       if (external) {
  prisma:client         if (!v.d || v.e > Decimal2.maxE) {
  prisma:client           x.e = NaN;
  prisma:client           x.d = null;
  prisma:client         } else if (v.e < Decimal2.minE) {
  prisma:client           x.e = 0;
  prisma:client           x.d = [0];
  prisma:client         } else {
  prisma:client           x.e = v.e;
  prisma:client           x.d = v.d.slice();
  prisma:client         }
  prisma:client       } else {
  prisma:client         x.e = v.e;
  prisma:client         x.d = v.d ? v.d.slice() : v.d;
  prisma:client       }
  prisma:client       return;
  prisma:client     }
  prisma:client     t = typeof v;
  prisma:client     if (t === "number") {
  prisma:client       if (v === 0) {
  prisma:client         x.s = 1 / v < 0 ? -1 : 1;
  prisma:client         x.e = 0;
  prisma:client         x.d = [0];
  prisma:client         return;
  prisma:client       }
  prisma:client       if (v < 0) {
  prisma:client         v = -v;
  prisma:client         x.s = -1;
  prisma:client       } else {
  prisma:client         x.s = 1;
  prisma:client       }
  prisma:client       if (v === ~~v && v < 1e7) {
  prisma:client         for (e = 0, i2 = v; i2 >= 10; i2 /= 10)
  prisma:client           e++;
  prisma:client         if (external) {
  prisma:client           if (e > Decimal2.maxE) {
  prisma:client             x.e = NaN;
  prisma:client             x.d = null;
  prisma:client           } else if (e < Decimal2.minE) {
  prisma:client             x.e = 0;
  prisma:client             x.d = [0];
  prisma:client           } else {
  prisma:client             x.e = e;
  prisma:client             x.d = [v];
  prisma:client           }
  prisma:client         } else {
  prisma:client           x.e = e;
  prisma:client           x.d = [v];
  prisma:client         }
  prisma:client         return;
  prisma:client       } else if (v * 0 !== 0) {
  prisma:client         if (!v)
  prisma:client           x.s = NaN;
  prisma:client         x.e = NaN;
  prisma:client         x.d = null;
  prisma:client         return;
  prisma:client       }
  prisma:client       return parseDecimal(x, v.toString());
  prisma:client     } else if (t !== "string") {
  prisma:client       throw Error(invalidArgument + v);
  prisma:client     }
  prisma:client     if ((i2 = v.charCodeAt(0)) === 45) {
  prisma:client       v = v.slice(1);
  prisma:client       x.s = -1;
  prisma:client     } else {
  prisma:client       if (i2 === 43)
  prisma:client         v = v.slice(1);
  prisma:client       x.s = 1;
  prisma:client     }
  prisma:client     return isDecimal.test(v) ? parseDecimal(x, v) : parseOther(x, v);
  prisma:client   },
  prisma:client       s: 1,
  prisma:client       e: 2,
  prisma:client       d: [
  prisma:client         789
  prisma:client       ]
  prisma:client     }
  prisma:client   }
  prisma:client }) +0ms
  prisma:client Generated request: +1ms
  prisma:client query {
  prisma:client   findUniqueChild(where: {
  prisma:client     id: 789
  prisma:client   }) {
  prisma:client     id
  prisma:client   }
  prisma:client }
  prisma:client  +0ms
  prisma:engine stdout Fetched a connection from the pool +6ms
  prisma:engine stdout Unknown error +1ms
prisma:query SELECT `main`.`Child`.`id` FROM `main`.`Child` WHERE `main`.`Child`.`id` IN (?,?) LIMIT ? OFFSET ?
```

But then it when it is separating those results back it's failing and the call to findUnique is returning null.

**Comment by tarazena** [comment]
Hey @mikeburgh, I took a look into your repo, here is how I got the resolvers to work with Decimal
in your `schema.js`
```js
const { objectType, queryField, list } = require('nexus');

exports.Child = objectType({
	name: 'Child',
	definition(t) {
		t.nonNull.string('id');
	},
});

exports.Parent = objectType({
	name: 'Parent',
	definition(t) {
		t.nonNull.string('id');
		t.nonNull.field('child', {
			type: 'Child',
			resolve: async (parent, args, ctx) => {
        return ctx.prisma.child.findFirst({
					where: {
						id: parent.childId.toNumber(),
					},
				});
			},
		});
	},
});

exports.ParentQuery = queryField('parents', {
	type: list('Parent'),
	resolve: (parent, args, ctx) => {
		return ctx.prisma.parent.findMany();
	},
});

```
the trick was with `findFirst` rather than `findUnique` and adding `toNumber()` to the `childId`

**Comment by mikeburgh** [comment]
@tarazena thanks for that, and yes that works, but it introduces the n+1 problem when you change from findUnique to findFirst.. here is the query logs from running that schema

```
prisma:query SELECT `main`.`Parent`.`id`, `main`.`Parent`.`childId` FROM `main`.`Parent` WHERE 1=1 LIMIT ? OFFSET ?
prisma:query SELECT `main`.`Child`.`id` FROM `main`.`Child` WHERE `main`.`Child`.`id` = ? LIMIT ? OFFSET ?
prisma:query SELECT `main`.`Child`.`id` FROM `main`.`Child` WHERE `main`.`Child`.`id` = ? LIMIT ? OFFSET ?
```

See the two calls to the child table in the above vs in the post above it makes one and uses the IN operator... that's the functionality that Prisma does which solves the N+1... but it's failing on Decimals when it comes to deconstructing the results from the single query.

**Comment by pantharshit00** [comment]
Hey @mikeburgh 

I tried your reproduction and I wasn't able to reproduce the N+1 I was able to reproduce another issue which is already reported:

![image](https://user-images.githubusercontent.com/22195362/110642037-a371df00-81d8-11eb-97c2-328913e60c7e.png)


But we have correct calls and data in the database but still client is returning null here. 

SQL is correctly batched:
![image](https://user-images.githubusercontent.com/22195362/110642117-b7b5dc00-81d8-11eb-92bb-f9cbce5ad7a3.png)

So this is related to https://github.com/prisma/prisma/issues/5941 which we need to fix. So I will considered this blocked by that issue for now. Changing calls to findFirst won't solve N+1 as they are not batched.

**Comment by mikeburgh** [comment]
@pantharshit00 Thanks for looking into it.. and yes sorry about the bad description.. it's not an N+1 bug specifically. I will re test when #5941 is fixed and close this if that fixes it.

**Comment by pantharshit00** [comment]
Looks like I am still getting unresolved values for the field in 2.21.2 so #5941 didn't really addressed that. 

I uploaded the code here: https://github.com/harshit-test-org/prisma-issue-59552

We will need to fix that now.

**Comment by janpio** [comment]
Internal note: Had to be reopened as https://github.com/prisma/prisma/pull/15526 showed two tests were not actually passing: https://github.com/prisma/prisma/pull/15526/commits/3098dc1a3bf1383ff7b3450eb6c65eb034faf6d0
